src/brogue/Monsters.c

```diff
@@ -230,6 +230,30 @@ boolean canDirectlySeeMonster(creature *monst) {
     return false;
 }
 
+// Centralize the logic of whether an attacker can/will attack the defender
+boolean canAttack( creature *attacker, creature *defender ) {
+    if (attacker == NULL || defender == NULL) return false;
+
+    // Must be enemies
+    boolean areEnemies = monsterWillAttackTarget(attacker, defender);
+
+    // Must be able to attack the cell where the defender is located
+    boolean canAttackCell = !cellHasTerrainFlag(defender->loc.x, defender->loc.y, T_OBSTRUCTS_PASSABILITY)
+                    || (defender->info.flags & MONST_ATTACKABLE_THRU_WALLS);
+
+    // Must be able to detect the defender
+    boolean locationKnown = !monsterIsHidden(defender, attacker);
+
+    // If the player is the attacker, also check vision and telepathy
+    if (attacker == &player) {
+        locationKnown = (locationKnown && playerCanSee(defender->loc.x, defender->loc.y))
+                        || monsterRevealed(defender);
```

Comment:
This is equiv to canSeeMonster, right? Thoughts on using that?

Reply:
It's not equivalent actually - it's grouped differently. It's equivalent to `canSee || revealed` (which does mean location known) because the other `|| revealed` in `canSee` is made redundant by the outer one.